### PR TITLE
[lsp] fixed Document.applyEdit for multiple textEdits

### LINKED ISSUE #1156: [lsp] IOOBE in Document.applyEdit
Steps to reproduce:
- in Theia, delete multiple characters at the end of a line using backspace
- trigger undo
=> an IOOBE is thrown and swallowed in Document.applyEdit

### PR DESCRIPTION
**Pull request by JanKoehnlein**
Fixes #1156

### CODE REVIEW

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend [hunk 1]
```diff
@@ -119,31 +119,36 @@ import org.eclipse.lsp4j.Range
...
+                } else {
+                    val start = currentDocument.getOffSet(change.range.start)
+                    val end = currentDocument.getOffSet(change.range.end)
+                    currentDocument.contents.substring(0, start) + change.newText + currentDocument.contents.substring(end)
+                }
+            val newVersion = if (currentDocument.version !== null)
+                    currentDocument.version + 1
+                else
+                    null
+            currentDocument = new Document(newVersion, newContent)
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend]: Here we are forgetting the value of printSourceOnError

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend [hunk 2]
```diff
@@ -119,31 +119,36 @@ import org.eclipse.lsp4j.Range
...
     def String getSubstring(Range range) {
-    		val start = getOffSet(range.start)
-    		val end = getOffSet(range.end)
-    		return this.contents.substring(start, end)
+          val start = getOffSet(range.start)
+          val end = getOffSet(range.end)
+          return this.contents.substring(start, end)
     }
 
     def Document applyChanges(Iterable<? extends TextEdit> changes) {
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend]: Could you please add documentation that explicitly states the relation of the offsets / ranges in the iterable of TextEdits, e.g. that the second text edit has a range that relates to the result after applying the first edit?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend [hunk 3]
```diff
@@ -119,31 +119,36 @@ import org.eclipse.lsp4j.Range
...
+        var currentDocument = this
         for (change : changes) {
-            if (change.range === null) {
-                newContent = change.newText
-            } else {
-                val start = getOffSet(change.range.start)
-                val end = getOffSet(change.range.end)
-                newContent = newContent.substring(0, start) + change.newText + newContent.substring(end)
-            }
+            val newContent = if (change.range === null) {
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend]: I wonder if this is correct, given the spec of the lsp:
```
TextEdit[]
Complex text manipulations are described with an array of TextEdit’s, representing a single change to the document.

All text edits ranges refer to positions in the original document. Text edits ranges must never overlap, that means no part of the original document must be manipulated by more than one edit. However, it is possible that multiple edits have the same start position: multiple inserts, or any number of inserts followed by a single remove or replace edit. If multiple inserts have the same position, the order in the array defines the order in which the inserted strings appear in the resulting text.
```

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend (hunk 1 above)
Comment (szarnekow) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend]: Also it appears the we should create only a single new version (see above)

**Comment by cdietrich** [comment]
i wonder how the requests look like that theia sends

**Comment by JanKoehnlein** [comment]
> I've the impression that the fix is not in line with the spec of LSP.

Maybe not with the spec. I've tried in VSCode and I get exactly the same requests and behavior as in Theia. So it seems to be Monaco, which is kind of the reference client for LSP.

**Comment by JanKoehnlein** [comment]
> i wonder how the requests look like that theia sends

Theia (and VSCode) send a didChangeRequest for the undo with individual TextEdits for each character that have increasing character positions, like in the test in this commit.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend (hunk 1 above)
Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend]: Fixed

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend (hunk 2 above)
Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend]: Done

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend]: TextDocumentContentChangeEvent do seem different from TextEdits.
i find it bogus that the semantics is different here

**Comment by JanKoehnlein** [comment]
I am not sure what to do here. Seems that LSP's spec and VS Code/Monaco behave differently here.
Adding @akosyakov to the conversation.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend]: => @JanKoehnlein do you know why we use the same code for text edits and contentChanges

**Comment by JanKoehnlein** [comment]
BTW, thanks a lot for the fast reviews.

**Comment by cdietrich** [comment]
as said if the comment i think the problem is that the specs differs between text edits and TextDocumentContentChangeEvent
but we use the same code to apply them

**Comment by cdietrich** [comment]
here is what vscode sends

undo
```
{
  "jsonrpc": "2.0",
  "method": "textDocument/didChange",
  "params": {
    "textDocument": {
      "version": 5,
      "uri": "file:///data/git/xtext-languageserver-example/demo/b.mydsl"
    },
    "contentChanges": [
      {
        "range": {
          "start": {
            "line": 1,
            "character": 23
          },
          "end": {
            "line": 1,
            "character": 23
          }
        },
        "rangeLength": 0,
        "text": "x"
      },
      {
        "range": {
          "start": {
            "line": 1,
            "character": 24
          },
          "end": {
            "line": 1,
            "character": 24
          }
        },
        "rangeLength": 0,
        "text": "x"
      }
    ]
  }
}
```

typing

```
{
  "jsonrpc": "2.0",
  "method": "textDocument/didChange",
  "params": {
    "textDocument": {
      "version": 6,
      "uri": "file:///data/git/xtext-languageserver-example/demo/b.mydsl"
    },
    "contentChanges": [
      {
        "range": {
          "start": {
            "line": 1,
            "character": 25
          },
          "end": {
            "line": 1,
            "character": 25
          }
        },
        "rangeLength": 0,
        "text": "y"
      }
    ]
  }
}
```

dont know if there are text edits sent/applied on server side (with the different semantics)

**Comment by JanKoehnlein** [comment]
@cdietrich I see. I'll revert and add a separate impl then for `TextDocumentContentChangeEvent`

**Comment by cdietrich** [comment]
yes it seems the conversion we do e.g. at org.eclipse.xtext.ide.server.LanguageServerImpl.didChange(DidChangeTextDocumentParams)
is a bit optimistic. 
i dont know if there are other places we do it too

**Comment by JanKoehnlein** [comment]
The semantics of `TextDocumentChangeEvent[]` (see https://github.com/microsoft/vscode/issues/23173#issuecomment-289378160) is different from `TextEdit[]` (see LSP spec as cited by @szarnekow).

We should not have implemented `Document#applyEdit(TextEdit[])` cause the server never uses `TextEdits`. We only do in tests. Anyway, it's API now, so I leave it and add a new method.

**Comment by cdietrich** [comment]
please also mark the old method as deprecated

**Comment by JanKoehnlein** [comment]
Great! Thanks a lot for your reviews and constructive comments.
Any idea why the ECA check is pending? Am I allowed to merge anyway?

**Comment by cdietrich** [comment]
you have to hit the revalidate button

**Pull request merged**